packages/react-core/src/components/Tooltip/Tooltip.tsx

```diff
@@ -50,150 +49,183 @@ export interface TooltipProps {
+      } else if (event.keyCode === KEY_CODES.ENTER && !visible) {
+        show();
+      }
+    }
   };
-
-  handleEscKeyClick = (event: KeyboardEvent) => {
-    if (event.keyCode === KEY_CODES.ESCAPE_KEY && this.tip.state.isVisible) {
-      this.tip.hide();
+  React.useEffect(() => {
+    if (triggerManually) {
+      if (isVisible) {
+        show();
+      } else {
+        hide();
+      }
+    }
+  }, [isVisible, triggerManually]);
+  const show = () => {
+    if (transitionTimerRef.current) {
+      clearTimeout(transitionTimerRef.current);
+    }
+    if (hideTimerRef.current) {
+      clearTimeout(hideTimerRef.current);
     }
+    showTimerRef.current = setTimeout(() => {
+      setVisible(true);
+      setOpacity(1);
+    }, entryDelay);
   };
+  const hide = () => {
+    if (showTimerRef.current) {
+      clearTimeout(showTimerRef.current);
+    }
+    hideTimerRef.current = setTimeout(() => {
+      setOpacity(0);
+      transitionTimerRef.current = setTimeout(() => setVisible(false), animationDuration);
+    }, exitDelay);
+  };
+  const positionModifiers = {
+    top: styles.modifiers.top,
+    bottom: styles.modifiers.bottom,
+    left: styles.modifiers.left,
+    right: styles.modifiers.right
+  };
+  const hasCustomMaxWidth = maxWidth !== tooltipMaxWidth.value;
+  const content = (
+    <div
+      className={css(styles.tooltip, className)}
+      role="tooltip"
+      id={id}
+      style={{
+        maxWidth: hasCustomMaxWidth ? maxWidth : null,
+        opacity,
+        transition: `opacity ${animationDuration}ms cubic-bezier(.54, 1.5, .38, 1.11)`
+      }}
+      {...rest}
+    >
+      <TooltipArrow />
+      <TooltipContent isLeftAligned={isContentLeftAligned}>{bodyContent}</TooltipContent>
+    </div>
+  );
 
-  componentDidMount() {
-    document.addEventListener('keydown', this.handleEscKeyClick, false);
-  }
-
-  componentWillUnmount() {
-    document.removeEventListener('keydown', this.handleEscKeyClick, false);
-  }
+  const onDocumentClick = (event: MouseEvent, triggerElement: HTMLElement) => {
+    // event.currentTarget = document
+    // event.target could be triggerElement or something else
+    if (hideOnClick === true) {
+      // hide on inside the toggle as well as on outside clicks
+      if (visible) {
+        hide();
+      } else if (event.target === triggerElement) {
+        show();
+      }
+    } else if (hideOnClick === 'toggle' && event.target === triggerElement) {
+      // prevent outside clicks from hiding but allow it to still be toggled on toggle click
+      if (visible) {
+        hide();
+      } else {
+        show();
+      }
+    } else if (hideOnClick === false && !visible && event.target === triggerElement) {
+      show();
+    }
+  };
 
-  extendChildren() {
-    return React.cloneElement(this.props.children, {
-      isAppLauncher: this.props.isAppLauncher
-    });
-  }
+  const addAriaToTrigger = () => {
+    if (aria === 'describedby' && !children.props['aria-describedby']) {
```

Comment:
Also needs `&& children.props` since `children` can be a `string`.

On that note, are strings fully supported or should the type be changed to `children: React.ReactElement`?

Reply:
should be ReactElement, good call!

Reply:
oh, looks like that is already the case! `children: ReactElement<any>;`